Make UI more like the go tool: add build cmd and -c and -o options

FiloSottile opened the pull request:
Implements the design in #30

See #8

Replaces #27

FiloSottile commented:
Small bug to fix before merging: godebug test foo/... will also try to include packages like foo/bar_test. It probably needs a different way to do expansion.

FiloSottile commented on cmd.go:
```diff
@@ -487,8 +579,24 @@ func doOutput(args []string) {
+func parseBuildArguments(args []string) ([]string, bool) {
+	if len(args) == 0 {
+		return []string{"."}, true
+		// logFatal("relative packages are not supported, specify what to build")
```
TODO: remove this leftover

jeremyschlatter commented:
This is great, thanks. Looking through it now.

jeremyschlatter commented on cmd.go:
```diff
@@ -77,30 +77,54 @@ must not be relative.
-
-See also: 'go help testflag'.
-`)
+See also: 'go help testflag'. Not that you have to use the 'test.'
```
Not -> Note

jeremyschlatter commented on cmd.go:
```diff
@@ -487,8 +597,23 @@ func doOutput(args []string) {
+	return args, false
+}
+
+func parseTestArguments(args []string) ([]string, []string) {
```
I find the named return parameters to be helpful documentation here. Was there a reason you removed them?

FiloSottile replied:
I have a personal allergy for the code that they encourage to write, but true, documentation comes first. Refactoring them back in.

jeremyschlatter commented on cmd.go:
```diff
@@ -212,28 +276,58 @@ func doTest(args []string) {
+	shellGo("", []string{"test", "-tags", *tags, "-i"}, packages)
+
+	// The target binary goes to -o if specified, otherwise to the default name
+	// is -c is specified, otherwise to the temporary directory.
```
is -> if

jeremyschlatter commented on cmd.go:
```diff
@@ -498,14 +623,17 @@ func parseTestArguments(args []string) (packages, testFlags []string) {
+			!strings.HasPrefix(arg, "-godebugwork") &&
+			!strings.HasPrefix(arg, "-tags") &&
+			!strings.HasPrefix(arg, "-o") &&
+			!strings.HasPrefix(arg, "-c") {
```
The `HasPrefix` check breaks if the user tries to pass a flag to their test that shares the same prefix. This was ok for "instrument" and "godebugwork", but there is a pretty good chance of conflict with "o" and "c".

Something like this would be better:

```go
isFlag := func(arg, name string) bool {
    return arg == name || strings.HasPrefix(arg, name + "=")
}
sep := len(args)
for i, arg := range args {
    if strings.HasPrefix(arg, "--") {
        arg = arg[1:]
    }
    if strings.HasPrefix(arg, "-") &&
        !isFlag(arg, "-instrument") &&
        !isFlag(arg, "-godebugwork") &&
        !isFlag(arg, "-tags") &&
        !isFlag(arg, "-o") &&
        !isFlag(arg, "-c") &&
```

FiloSottile replied:
Good catch, thanks

jeremyschlatter commented on testdata/test-filesystem/gopath/src/tags/tagged.go:
```diff
@@ -0,0 +1,7 @@
+
+func init() {
+	v = "y"
+}
```
Thanks for adding test files. I don't see test cases that use them, though. Can you add those?

FiloSottile commented:
Should have addressed all the points, except the tests. I haven't had the time to figure out how the tests scaffolding works (I added the files while I was testing manually).

There are various aspects of the build and test commands that need testing, too, but that will require some more work to make the tests run the binary etc.

jeremyschlatter commented:
Looks great, thanks.
